### PR TITLE
chore: add options for resource requests for vault-agent

### PR DESCRIPTION
**Pull request by dhumphries-sainsburys**
Currently the vault-agent takes a resource for requests and limits combined but in usage we see that startup usage is high and ongoing usage is low so we are faced with either large amounts of wasted resources across a cluster or a risk of OOM kills on startup. This PR will just allow requests and limits to be defined independently of each other

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -46,7 +46,7 @@ build: ## Build binary
 
 .PHONY: container-image
 container-image: ## Build container image
-	docker build -t ${CONTAINER_IMAGE_REF} .
+	docker build -t ${CONTAINER_IMAGE_REF} . --platform linux/amd64
```
Comment (ramizpolic) [inline Makefile:49]: I don't think we need to specify a target platform here. If you would like to do that, you will also need to update the github actions to build and release those images too.

File: go.mod [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module github.com/bank-vaults/vault-secrets-webhook
 
-go 1.21.1
+go 1.21
```
Comment (ramizpolic) [inline go.mod:3]: `go 1.21` is not a valid go toolchain/tag (https://github.com/golang/go/tags); `go 1.21.0` is

File: pkg/webhook/config.go [hunk 3]
```diff
@@ -318,16 +320,28 @@ func parseVaultConfig(obj metav1.Object, ar *model.AdmissionReview) VaultConfig
...
 	}
 
-	if val, err := resource.ParseQuantity(annotations["vault.security.banzaicloud.io/vault-agent-memory"]); err == nil {
-		vaultConfig.AgentMemory = val
+	if val, err := resource.ParseQuantity(annotations["vault.security.banzaicloud.io/vault-agent-memory-limit"]); err == nil {
+		vaultConfig.AgentMemoryLimit = val
 	} else {
-		vaultConfig.AgentMemory = resource.MustParse("128Mi")
+		vaultConfig.AgentMemoryLimit = resource.MustParse("128Mi")
+	}
```
Comment (ramizpolic) [inline pkg/webhook/config.go:333]: This is going to break compatibility, can we use the `vault.security.banzaicloud.io/vault-agent-memory` as main and `vault.security.banzaicloud.io/vault-agent-memory-limit` as a backup option, e.g.
```go
if val, err := resource.ParseQuantity(annotations["vault.security.banzaicloud.io/vault-agent-memory"]); err == nil {
	vaultConfig.AgentMemoryLimit = val
} else if val, err := resource.ParseQuantity(annotations["vault.security.banzaicloud.io/vault-agent-memory-limit"]); err == nil {
	vaultConfig.AgentMemoryLimit = val
} else {
	vaultConfig.AgentMemoryLimit = resource.MustParse("128Mi")
}
```

**Review by ramizpolic — CHANGES_REQUESTED** [review]

**Comment by ramizpolic** [comment]
Hi @dhumphries-sainsburys, thanks for the contribution! I left a couple of notes to address and we can merge it afterwards.

File: Makefile (hunk 1 above)
Comment (dhumphries-sainsburys) [reply Makefile:49]: This was just something i needed to do to build it locally as i am using a mac

Comment (ramizpolic) [reply Makefile:49]: You can use env to set the default platform via `export DOCKER_DEFAULT_PLATFORM=linux/amd64`

File: pkg/webhook/config.go [hunk 4]
```diff
@@ -318,16 +320,38 @@ func parseVaultConfig(obj metav1.Object, ar *model.AdmissionReview) VaultConfig
...
 	}
 
-	if val, err := resource.ParseQuantity(annotations["vault.security.banzaicloud.io/vault-agent-cpu"]); err == nil {
-		vaultConfig.AgentCPU = val
+	if val, err := resource.ParseQuantity(annotations["vault.security.banzaicloud.io/vault-agent-cpu-limit"]); err == nil {
+		vaultConfig.AgentCPULimit = val
 	} else {
-		vaultConfig.AgentCPU = resource.MustParse("100m")
+		vaultConfig.AgentCPULimit = resource.MustParse("100m")
 	}
```
Comment (ramizpolic) [inline pkg/webhook/config.go:327]: These are extra as they are already added below (line 329 to 335).

File: pkg/webhook/config.go [hunk 5]
```diff
@@ -318,16 +320,38 @@ func parseVaultConfig(obj metav1.Object, ar *model.AdmissionReview) VaultConfig
...
 
-	if val, err := resource.ParseQuantity(annotations["vault.security.banzaicloud.io/vault-agent-memory"]); err == nil {
-		vaultConfig.AgentMemory = val
+    if val, err := resource.ParseQuantity(annotations["vault.security.banzaicloud.io/vault-agent-cpu"]); err == nil {
```
Comment (ramizpolic) [inline pkg/webhook/config.go:329]: Could you just leave a comment on this one that it is preserved for backward compatibility? The same for the memory label. The same should be reflected in the docs PR.

File: Makefile (hunk 1 above)
Comment (ramizpolic) [reply Makefile:49]: You can remove this from the makefile

File: go.mod (hunk 2 above)
Comment (ramizpolic) [inline go.mod:3]: no need to change this

**Review by ramizpolic — CHANGES_REQUESTED** [review]
Added a couple of notes, can merge after these are resolved. Thanks for the updates @dhumphries-sainsburys!

File: go.mod (hunk 2 above)
Comment (dhumphries-sainsburys) [reply go.mod:3]: It threw errors about 1.21.1 not being a thing when i built it

**Comment by dhumphries-sainsburys** [comment]
@ramizpolic - Think all changes are made but github UI seems to be showing something different to the commits so not entirely sure

**Comment by dhumphries-sainsburys** [comment]
Squashed the commits as well as signed. In regards to the linter i couldn't find a problem and the error previous looked more like a problem with the linter rather than what it was linting to me

**Review by ramizpolic — APPROVED** [review]
great work @dhumphries-sainsburys, thanks for the contribution!

**Pull request merged**